system/lib/wasmfs/js_impl_backend.h

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the JS file backend and JS file of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "backend.h"
+#include "wasmfs.h"
+
+//
+// A JS Impl backend has files that are implemented by JS code. Each backend
+// object in C++ can call JS to define the proper JS code on that side, and we
+// keep a mapping of C++ pointer of backend => JS code in JS. Then, when we do
+// something like a read from a JSImplFile we pass it the backend, find the
+// proper code to run, and run it. This adds a layer of indirection at the JS
+// level that makes it easy to write new backends in 99% JS.
+//
+// Each file operation in the _wasmfs_jsimpl_* APIs that we call from here take
+// the backend and a pointer to this file itself. Those allow the JS to identify
+// both the backend and the particular file. TODO: We could use dense indexes
+// instead of pointers, and use an array instead of a map.
+//
+// To write a new backend in JS, you basically do the following:
+//
+//  1. Add a declaration of the C function to create the backend in the
+//     "backend creation" section of emscripten/wasmfs.h.
+//  2. Add a cpp file for the new backend, and implement the C function from 1,
+//     which should create it on both the C++ (using JSImplBackend) and JS
+//     sides.
```

Comment:
The C layer should be a thin wrapper around the underlying C++ and should not perform any important initialization itself.

We can also improve this workflow so that JS backends can be defined completely in JS userspace with no need to modify the WasmFS source. All we need is a JS API function that creates an instance of the JSImplBackend and allows users to install methods on it and perform whatever other instantiation they want, all from JS. The `JSImplBackend` constructor should probably install the backend on the JS side then call a user-installable JS function to perform custom initialization.

Adding a small C wrapper to emscripten/wasmfs.h can be a step on top of that workflow for Emscripten developers creating "official" backends, so I would expect to see that documented as an optional last step rather than as the first step.

Reply:
> The C layer should be a thin wrapper around the underlying C++ and should not perform any important initialization itself.

Added a comment now, thanks.

> We can also improve this workflow so that JS backends can be defined completely in JS userspace with no need to modify the WasmFS source.

Yes, I think that can work. But it does mean a forced export which increases code size, both of the backend-generating code itself, and also the JSImplBackend itself. Those would end up bundled in all builds (unless metadce gets lucky and manages to remove parts). But I agree there are nice benefits to a pure JS approach. Perhaps that should be an option, say if the full JS API is enabled (`FORCE_FILESYSTEM`) - if we go that route, then user JS backends would need no C but perhaps our own would still have C files (so that they work without `FORCE_FILESYSTEM`).

Reply:
> Perhaps that should be an option, say if the full JS API is enabled (`FORCE_FILESYSTEM`) - if we go that route, then user JS backends would need no C but perhaps our own would still have C files (so that they work without `FORCE_FILESYSTEM`).

That makes sense to me.